### PR TITLE
TouchUp

### PR DESCRIPTION
**Pull request by markchristopherwest**
Includes initial support for:
- Boundary
- Waypoint

Includes API calls to Vault internal-counters (Client Count) in Vault v1.11.0 or later:
https://www.vaultproject.io/api-docs/system/internal-counters#client-count

### CODE REVIEW

**Comment by mkcp** [comment]
Thanks for the contribution.

Adding Boundary and Waypoint is part of our long term goals but we'll have to discuss more with their Support Engineering teams before we're ready to make experimental support available. hcdiag's product defaults are optimized for diagnosing and debugging support issues, so we must to ensure the API calls added here don't unnecessarily reach into products' application space and expose sensitive customer data. We want to avoid introducing security issues and stay aligned with hcdiag's data frugality goals.

For merge expectations: we're going to keep this PR around for a bit because we will likely accept the Boundary and Waypoint implementations in the `agent`, `client`, and `product` pkgs. The product-seekers as implemented will need further review. Complicating timing: we're currently making major revisions to our seeker runtime for `v0.4.0`, so expect some merge conflicts in this branch after then.

To ensure we can accept PRs quickly and cleanly, in the future we urge you to avoid submitting PRs with undescribed changes to the data collected like those in Terraform Enterprise. If customers wish to opt-in to these collections now, each of these seekers can be declared in HCL and provided via the `-config` flag.

File: product/waypoint.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+		Seekers: seekers,
+		Config:  cfg,
+	}, nil
+}
+
+// WaypointSeekers seek information about Waypoint.
+func WaypointSeekers(cfg Config, api *client.APIClient) ([]*s.Seeker, error) {
+	seekers := []*s.Seeker{
+		s.NewCommander("waypoint version", "string"),
+		s.NewCommander(fmt.Sprintf("waypoint debug -output=%s/WaypointDebug -duration=%s -interval=%s", cfg.TmpDir, cfg.DebugDuration, cfg.DebugInterval), "string"),
```
Comment (briancain) [inline product/waypoint.go:42]: FYI - This command doesn't exist in `waypoint`.

File: product/waypoint.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+		s.NewHTTPer(api, "/v1/agent/self"),
+		s.NewHTTPer(api, "/v1/agent/metrics"),
+		s.NewHTTPer(api, "/v1/catalog/datacenters"),
+		s.NewHTTPer(api, "/v1/catalog/services"),
+		s.NewHTTPer(api, "/v1/namespace"),
+		s.NewHTTPer(api, "/v1/status/leader"),
+		s.NewHTTPer(api, "/v1/status/peers"),
```
Comment (briancain) [inline product/waypoint.go:50]: These HTTP endpoints don't exist in Waypoint, but maybe they do in HCP Waypoint?

**Comment by PPacent** [comment]
@markchristopherwest - thank you for putting this together and thanks to @mkcp and @aliciaclark1066 for taking a look at this. 

Since Boundary doesn't have a self-managed enterprise distribution, there's not yet support engineers that would be using hcdiag to actively support customer Boundary environments. With this is mind, it seems like the primary benefit of adding hcdiag support for Boundary would be streamlining how the engineering team gathers environment configuration info when we invariably do adhoc community support of OSS environments. If that's the case, then I'm not sure it's yet worth our team taking on the support of hcdiag support for Boundary and want to be clear we won't be able to commit Boundary engineering resources at this time.

**Comment by aliciaclark1066** [comment]
Hey @markchristopherwest,

The Boundary and Waypoint teams looked at the PR and decided they aren't ready to support requests coming from hcdiag. We are also unable to accept the other changes regarding modifications to the Terraform defaults since Support has different requirements. We'll need to close the PR at this time without merging, but let's keep in touch!

CC - @mkcp

**Comment by markchristopherwest** [comment]
Perhaps @aliciaclark1066 however, I believe we need to revisit hcdiag support for both OSS & HCP for Boundary & Waypoint:

Even before GA, Boundary's got Swagger:

https://github.com/hashicorp/boundary/blob/main/internal/gen/controller.swagger.json

Without looking at Asana, mabe Evan & the Waypoint team are also working on Swagger: 

https://github.com/hashicorp/waypoint/search?q=swagger

Ran into both @evanphx & @jefferai as I was heading out of HashiConf & pretty sure can get their blessing.

Also, I am reopening this to include this gist of HCL to gather renewal metrics for FieldOps:

https://gist.github.com/markchristopherwest/b7b848e639364574c71e1051c5ba430b

Will reach out & schedule triage between SEA SME's for confirmation but I am thinking EOY.

Miss you @ HashiConf & have a happy holiday if I do not see you before!

**Comment by cotarg** [comment]
Hi @markchristoperwest, thanks for reaching out to follow up on this after your discussions at HashiConf. 

I want to take a moment to thank you for your enthusiasm around the tool! We're so happy as a team that you both appreciate the value of it, and submit possible improvements you see for integration. 

Our roadmap for hcdiag has pivoted since the last time we've chatted with you, and while I have some good news for you on that front...I do have to begin by apologizing that we will not be accepting this PR. I empathize that closing it again may cause some frustration, so I wanted to offer a little transparency on future plans!

We have plans to integrate all of the HashiCorp product offerings across Cloud, Enterprise, and OSS, however before we can do that we have some refactor and core functionality to finish and polish. We're prioritizing finishing this refactor work over the next couple of quarters in order to enable us to smoothly extend hcdiag in the future.

We'd love to chat with you more when we're ready to talk about integrating Boundary and Waypoint (or any other product you'd like to share with us about)! @aliciaclark1066 has you on her list of folks to make contact with, and we'd also love to have your input at any of the information sessions we host moving forward! This sort of enthusiasm for our tool is truly appreciated, and we'd love to harness it to make our product better when we're ready to look at new integrations!

Thanks again, and please feel free to reach out to me if you have any other questions!

**Pull request closed without merging**